File: packages/sdks/flagmind-ruby/lib/flagmind/rule_matcher.rb
require "set"

module Tombstone
  module RuleMatcher
    GEO_ATTRIBUTES = Set.new(["geo.country", "geo.region"])

    # Canonical model: dot-notation attribute resolution over a flat attrs hash
    # (this release's EvaluationContext.attrs is Hash, so multi-segment paths
    # like "geo.country" resolve via nested-map convention where the caller
    # stores nested structures). Returns nil if the attribute is not present.
    def self.resolve_attribute(attribute, context)
      return context.user_id if attribute == "user_id"
      return context.org_id if attribute == "org_id"

      # Dot-notation resolution: split on dots, traverse nested hashes
      segments = attribute.split(".")
      current = context.attrs
      segments.each do |seg|
        return nil unless current.is_a?(Hash) && current.key?(seg)
        current = current[seg]
      end

      # Fallback: if only one segment and it's a flat key, return it
      if segments.size == 1 && context.attrs.key?(attribute)
        return context.attrs[attribute]
      end

      current
    end

    def self.evaluate_condition(condition, context)
      raw = resolve_attribute(condition.attribute, context)
      raise InconclusiveMatchError, "Attribute '#{condition.attribute}' not present in evaluation context" if raw.nil?

      attr_val = raw.to_s
      op = normalize_operator(condition.operator)
      values = condition.values
      is_geo = GEO_ATTRIBUTES.include?(condition.attribute)

      result = case op
      when "eq", "in"
        is_geo ? contains_ignore_case(values, attr_val) : values.include?(attr_val)
      when "neq", "nin"
        is_geo ? !contains_ignore_case(values, attr_val) : !values.include?(attr_val)
      when "contains"
        any_contains_ignore_case(values, attr_val)
      when "startswith"
        any_starts_with_ignore_case(values, attr_val)
      when "endswith"
        any_ends_with_ignore_case(values, attr_val)
      when "gt", "gte", "lt", "lte"
        evaluate_numeric(op, attr_val, values, condition.attribute)
      else
        raise InconclusiveMatchError, "Unknown operator: '#{op}'"
      end

      condition.negate ? !result : result
    end

    def self.normalize_operator(operator)
      op = operator.downcase
      case op
      when "not_in" then "nin"
      when "prefix" then "startswith"
      when "suffix" then "endswith"
      else op
      end
    end

    def self.contains_ignore_case(values, attr_val)
      upper = attr_val.upcase
      values.any? { |v| v.to_s.upcase == upper }
    end

    def self.any_contains_ignore_case(values, attr_val)
      upper_attr = attr_val.upcase
      values.any? { |v| upper_attr.include?(v.to_s.upcase) }
    end

    def self.any_starts_with_ignore_case(values, attr_val)
      upper_attr = attr_val.upcase
      values.any? { |v| upper_attr.start_with?(v.to_s.upcase) }
    end

    def self.any_ends_with_ignore_case(values, attr_val)
      upper_attr = attr_val.upcase
      values.any? { |v| upper_attr.end_with?(v.to_s.upcase) }
    end

    def self.evaluate_numeric(op, attr_val, values, attribute)
      begin
        n_attr = Float(attr_val)
        n_val = Float(values[0])
      rescue ArgumentError, TypeError, IndexError
        raise InconclusiveMatchError, "Numeric cast failed for '#{attribute}'"
      end

      case op
      when "gt" then n_attr > n_val
      when "gte" then n_attr >= n_val
      when "lt" then n_attr < n_val
      when "lte" then n_attr <= n_val
      else false
      end
    end
  end
end

File: packages/sdks/flagmind-dotnet/src/FlagMind/RuleMatcher.cs
using System.Text.RegularExpressions;
using System.Globalization;

namespace Tombstone;

public static class RuleMatcher
{
    private static readonly HashSet<string> GeoAttributes = new() { "geo.country", "geo.region" };

    /// <summary>
    /// Canonical model: attribute resolution over the flat Attrs dictionary. This
    /// release's EvaluationContext.Attrs is Dictionary&lt;string,string&gt;, so
    /// multi-segment paths like "geo.country" resolve as literal flat keys — the
    /// wire format flattens nested JSON before populating Attrs, matching the
    /// same convention used by the Java/Ruby ports of this canonical model.
    /// </summary>
    public static object? ResolveAttribute(string attribute, EvaluationContext context)
    {
        if (attribute == "user_id") return context.UserId;
        if (attribute == "org_id") return context.OrgId;
        return context.AttrsOrEmpty.TryGetValue(attribute, out var value) ? value : null;
    }

    public static bool EvaluateCondition(PropertyCondition condition, EvaluationContext context)
    {
        var raw = ResolveAttribute(condition.Attribute, context);
        if (raw is null)
            throw new InconclusiveMatchException(
                $"Attribute '{condition.Attribute}' not present in evaluation context");

        var attrVal = raw.ToString() ?? "";
        var op = NormalizeOperator(condition.Operator);
        var values = condition.Values;
        var isGeo = GeoAttributes.Contains(condition.Attribute);

        bool result = op switch
        {
            "eq" or "in" => isGeo ? ContainsIgnoreCase(values, attrVal) : values.Contains(attrVal),
            "neq" or "nin" => isGeo ? !ContainsIgnoreCase(values, attrVal) : !values.Contains(attrVal),
            "contains" => AnyContainsIgnoreCase(values, attrVal),
            "startswith" => AnyStartsWithIgnoreCase(values, attrVal),
            "endswith" => AnyEndsWithIgnoreCase(values, attrVal),
            "gt" or "gte" or "lt" or "lte" => EvaluateNumeric(op, attrVal, values, condition.Attribute),
            "semver_gt" or "semver_gte" or "semver_lt" or "semver_lte" or "semver_eq"
                => EvaluateSemver(op, attrVal, values, condition.Attribute),
            "date_before" or "date_after"
                => EvaluateDate(op, attrVal, values, condition.Attribute),
            _ => throw new InconclusiveMatchException($"Unknown operator: '{op}'"),
        };

        return condition.Negate ? !result : result;
    }

    private static string NormalizeOperator(string operatorName)
    {
        var op = operatorName.ToLowerInvariant();
        return op switch
        {
            "not_in" => "nin",
            "prefix" => "startswith",
            "suffix" => "endswith",
            _ => op,
        };
    }

    private static bool ContainsIgnoreCase(List<string> values, string attrVal)
    {
        var upper = attrVal.ToUpperInvariant();
        return values.Any(v => v.ToUpperInvariant() == upper);
    }

    private static bool AnyContainsIgnoreCase(List<string> values, string attrVal)
    {
        var upperAttr = attrVal.ToUpperInvariant();
        return values.Any(v => upperAttr.Contains(v.ToUpperInvariant()));
    }

    private static bool AnyStartsWithIgnoreCase(List<string> values, string attrVal)
    {
        var upperAttr = attrVal.ToUpperInvariant();
        return values.Any(v => upperAttr.StartsWith(v.ToUpperInvariant(), StringComparison.Ordinal));
    }

    private static bool AnyEndsWithIgnoreCase(List<string> values, string attrVal)
    {
        var upperAttr = attrVal.ToUpperInvariant();
        return values.Any(v => upperAttr.EndsWith(v.ToUpperInvariant(), StringComparison.Ordinal));
    }

    // Uses invariant culture (not the current thread's culture) so decimal-separator
    // parsing is identical on every machine regardless of OS locale — matching
    // Java's Double.parseDouble and Ruby's Float(), which are always invariant.
    private static bool EvaluateNumeric(string op, string attrVal, List<string> values, string attribute)
    {
        var style = System.Globalization.NumberStyles.Float;
        var culture = System.Globalization.CultureInfo.InvariantCulture;
        if (values.Count == 0
            || !double.TryParse(attrVal, style, culture, out var nAttr)
            || !double.TryParse(values[0], style, culture, out var nVal))
            throw new InconclusiveMatchException($"Numeric cast failed for '{attribute}'");

        return op switch
        {
            "gt" => nAttr > nVal,
            "gte" => nAttr >= nVal,
            "lt" => nAttr < nVal,
            "lte" => nAttr <= nVal,
            _ => false,
        };
    }

    private static readonly Regex LeadingVOrBuildMetadata = new(@"(^v|\+.*$)", RegexOptions.Compiled);
    private static readonly Regex PureDigits = new(@"^\d+$", RegexOptions.Compiled);

    /// <summary>Ported byte-for-byte from flagmind-python's matching.py:27-39 (GrowthBook pattern).</summary>
    public static string PaddedVersion(string v)
    {
        v = LeadingVOrBuildMetadata.Replace(v, "");
        var parts = v.Split('-', '.');
        var padded = parts.Select(p => PureDigits.IsMatch(p) ? p.PadLeft(5, ' ') : p).ToList();
        if (padded.Count == 3) padded.Add("~");
        return string.Join(".", padded);
    }

    private static bool EvaluateSemver(string op, string attrVal, List<string> values, string attribute)
    {
        if (values.Count == 0)
            throw new InconclusiveMatchException($"semver operator requires at least one value for '{attribute}'");

        var a = PaddedVersion(attrVal);
        var b = PaddedVersion(values[0]);
        var cmp = string.CompareOrdinal(a, b);

        return op switch
        {
            "semver_gt" => cmp > 0,
            "semver_gte" => cmp >= 0,
            "semver_lt" => cmp < 0,
            "semver_lte" => cmp <= 0,
            "semver_eq" => cmp == 0,
            _ => false,
        };
    }

    private static bool EvaluateDate(string op, string attrVal, List<string> values, string attribute)
    {
        if (values.Count == 0
            || !DateTimeOffset.TryParse(NormalizeIso8601(attrVal), CultureInfo.InvariantCulture, DateTimeStyles.None, out var dtAttr)
            || !DateTimeOffset.TryParse(NormalizeIso8601(values[0]), CultureInfo.InvariantCulture, DateTimeStyles.None, out var dtVal))
            throw new InconclusiveMatchException($"Date parse failed for '{attribute}'");

        return op == "date_before" ? dtAttr < dtVal : dtAttr > dtVal;
    }

    private static string NormalizeIso8601(string s) => s.Replace("Z", "+00:00");
}
